pipeline/dags/download_images.py

```diff
@@ -0,0 +1,255 @@
+            If not, cancel pipeline.
+            """
+            api_ok = HttpSensor(
+                task_id="http_sensor", http_conn_id=HTTP_CONN_ID, endpoint="/"
+            ).poke(context={})
+
+            bindings = read_bindings(set_id)
+
+            if not bindings:
+                print("No new bindings after previous download.")
+                return "cancel_pipeline"
+            if not api_ok:
+                print("NLF api does not respond.")
+                return "cancel_pipeline"
+            else:
+                return "begin_download"
+
+        check_if_download_should_begin() >> [begin_download, cancel_pipeline]
+
+        @task_group(group_id="download_set")
+        def download_set(set_id, api, ssh_conn_id):
+
+            bindings = read_bindings(set_id)
+
+            if INITIAL_DOWNLOAD:
+                image_split = utils.assign_bindings_to_images(bindings, 150)
+                save_image_split(image_split, set_id)
+
+            else:
+                image_split = utils.assign_update_bindings_to_images(
+                    bindings, IMAGE_SPLIT_DIR / f"{set_id}_images.json"
+                )
+
+            image_downloads = []
+
+            for image in image_split:
+                if image["bindings"]:
+
+                    @task_group(group_id=f"download_image_{set_id}_{image['prefix']}")
+                    def download_image(image):
+                        @task(task_id=f"ensure_image_{set_id}_{image['prefix']}")
+                        def ensure_image(image):
+                            """
+                            Create an empty directory for image contents or extract an existing disk image.
+                            """
+
+                            image_base_name = f"{set_id}_{image['prefix']}".rstrip("_")
+                            image_dir_path = os.path.join(BASE_PATH, image_base_name)
+
+                            # Check if image exists
+                            ssh_hook = SSHHook(ssh_conn_id=ssh_conn_id)
+                            with ssh_hook.get_conn() as ssh_client:
+                                sftp_client = ssh_client.open_sftp()
+                                utils.make_intermediate_dirs(
+                                    sftp_client=sftp_client,
+                                    remote_directory=BASE_PATH,
+                                )
+                                # if exists, extract with a bash command
+                                if f"{image_base_name}.sqfs" in sftp_client.listdir(
+                                    BASE_PATH
+                                ):
+                                    sftp_client.chdir(BASE_PATH)
+                                    ssh_client.exec_command(
+                                        f"unsquashfs -d {image_dir_path} {image_dir_path}.sqfs"
+                                    )
+                                    ssh_client.exec_command(f"rm {image_dir_path}.sqfs")
```

Comment:
I don't think we have, actually. Since we are downloading one image at a time, space should not be an issue, so I would propose to move the deletion of the old image to just before the creation of the new one.